GiniVision.podspec

```diff
@@ -29,6 +29,7 @@ The Gini Vision Library for iOS provides functionality to capture documents with
     networking.source_files = 'GiniVision/Classes/Networking/*.swift', 'GiniVision/Classes/Networking/Extensions/*.swift'
     networking.dependency "GiniVision/Core"
     networking.dependency "Gini-iOS-SDK", "~> 1.2"
+    networking.dependency "Bolts", "~> 1.9"
```

Comment:
Currently, we are using the last release version 1.5.3. We know, there is a newer version available as rc1 and we will update soon.

Anyway, I think though, the dependency should be explicitly specified since GiniVision is a framework of its own, and thus, requires all its dependencies listed when linking. Otherwise, the linker option `-framework Bolts` is omitted from the parameters when ld is invoked. And despite the fact the library Bolts is there and a path to that libraries is also specified in the parameters when invoking ld for GiniVision, the linker ld will issue the error.

Reply:
Well, the _GiniVision_ library does not have a direct dependency to neither _Bolts_ nor _TrustKit_, since both are part of the _Gini-iOS-SDK_. As pointed out [here](https://guides.cocoapods.org/syntax/podspec.html#subspec), you can specify inner dependencies for subpec (which in this case makes sense).
Could you copy-paste your Podfile configuration? Maybe

Reply:
Could you give me more information about your Build settings (_Valid Architecture, Other Linker flag, Architectures..._)?

Reply:
Some GiniVision sources in folder `Networking` in fact directly reference symbols from the Bolts framework. ;)   You may find `BFTask` and `BFCancellationTokenSource` in these sources. 
Please see my updated PR where I explicitly added the import directive `import Bolts`.

The ARCH build setting is set to be the $(ARCHS_STANDARD), VALID_ARCHS equals arm64, arm64e, armv7, armv7s.

Reply:
As I said, `Bolts` is a dependency on its own of the `Gini-iOS-SDK` and there is no need to create specific references in the `GiniVision` library.

Could you verify that your `Podfile` look like this (it's important that the `Gini-iOS-SDK` doesn't appear in the list):
```
source 'https://github.com/gini/gini-podspecs.git'
source 'https://github.com/CocoaPods/Specs.git'
platform :ios, '9.0'

use_frameworks!

target 'Your_app Swift' do

    ....
    pod 'GiniVision'
    pod 'GiniVision/Networking'

end

....
```

Reply:
Using the pods as described above in our Podfile (and using CocoaPods Version 1.5.3), we get the following linker error:

```
Undefined symbols for architecture arm64:
  "_OBJC_CLASS_$_BFTask", referenced from:
      objc-class-ref in AccountingDocumentService.o
      objc-class-ref in DocumentService.o
      objc-class-ref in DocumentServiceProtocol.o
  "_OBJC_CLASS_$_BFCancellationTokenSource", referenced from:
      objc-class-ref in AccountingDocumentService.o
      objc-class-ref in DocumentService.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Note: the linker does not always issue this error. For example, when having Xcode open, performing pod install (or pod update), and then building - the error does not occur.
However, relaunching Xcode an then building again will issue the linker error. Which is a bit strange. IMHO, it should always issue the error ;) . We also use Xcode´s new build system.

We do no have an issue (yet) with a local fork, containing the changes in the PR. I will try CocoaPods 1.6 (it's released now) and update my comment.

Reply:
I don't think it's an issue with Cocoapods, since we have used it in several products and didn't have any issue. Regarding the new build system, we have been tracking an [issue](https://github.com/CocoaPods/CocoaPods/issues/8073#issuecomment-460286563) where the pods changes weren't detected in every build, having to use the old build system again. Either way, this shouldn't affect you since you are not using _Development pods_ in your project.

Could you quickly [create](https://guides.cocoapods.org/using/using-cocoapods.html) a new project and try to use the `Podfile` I mentioned above? This way we can rule out any issue in your Build settings.

Reply:
I created a Single View App, with the following Podfile:
```
source 'https://github.com/gini/gini-podspecs.git'
source 'https://github.com/CocoaPods/Specs.git'
platform :ios, '9.0'

target 'GiniDemo' do
    use_frameworks!

    pod 'AppCenter'
    pod 'Firebase/Core'
    pod 'Alamofire'
    pod 'GiniVision'
    pod 'GiniVision/Networking'
end
```

Not exactly a vanilla Gini demo, but the other Pods are currently used by our project and I thought it makes sense to add them to give a reasonable load to the build system.

We get the following error:
```

Showing All Messages
Ld /Users/agrosam/Library/Developer/Xcode/DerivedData/GiniDemo-dejipfpuyqllwcfijniaemxaqepo/Build/Products/Debug-iphonesimulator/GiniVision/GiniVision.framework/GiniVision normal x86_64 (in target: GiniVision)
    cd /Users/agrosam/Develop/GiniDemo/Pods
    export IPHONEOS_DEPLOYMENT_TARGET=9.0
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -arch x86_64 -dynamiclib -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator12.1.sdk -L/Users/agrosam/Library/Developer/Xcode/DerivedData/GiniDemo-dejipfpuyqllwcfijniaemxaqepo/Build/Products/Debug-iphonesimulator/GiniVision -F/Users/agrosam/Library/Developer/Xcode/DerivedData/GiniDemo-dejipfpuyqllwcfijniaemxaqepo/Build/Products/Debug-iphonesimulator/GiniVision -F/Users/agrosam/Library/Developer/Xcode/DerivedData/GiniDemo-dejipfpuyqllwcfijniaemxaqepo/Build/Products/Debug-iphonesimulator/Bolts -F/Users/agrosam/Library/Developer/Xcode/DerivedData/GiniDemo-dejipfpuyqllwcfijniaemxaqepo/Build/Products/Debug-iphonesimulator/Gini-iOS-SDK -filelist /Users/agrosam/Library/Developer/Xcode/DerivedData/GiniDemo-dejipfpuyqllwcfijniaemxaqepo/Build/Intermediates.noindex/Pods.build/Debug-iphonesimulator/GiniVision.build/Objects-normal/x86_64/GiniVision.LinkFileList -install_name @rpath/GiniVision.framework/GiniVision -Xlinker -rpath -Xlinker @executable_path/Frameworks -Xlinker -rpath -Xlinker @loader_path/Frameworks -mios-simulator-version-min=9.0 -dead_strip -Xlinker -object_path_lto -Xlinker /Users/agrosam/Library/Developer/Xcode/DerivedData/GiniDemo-dejipfpuyqllwcfijniaemxaqepo/Build/Intermediates.noindex/Pods.build/Debug-iphonesimulator/GiniVision.build/Objects-normal/x86_64/GiniVision_lto.o -Xlinker -export_dynamic -Xlinker -no_deduplicate -Xlinker -objc_abi_version -Xlinker 2 -fobjc-arc -fobjc-link-runtime -L/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/iphonesimulator -Xlinker -add_ast_path -Xlinker /Users/agrosam/Library/Developer/Xcode/DerivedData/GiniDemo-dejipfpuyqllwcfijniaemxaqepo/Build/Intermediates.noindex/Pods.build/Debug-iphonesimulator/GiniVision.build/Objects-normal/x86_64/GiniVision.swiftmodule -framework AVFoundation -framework CoreMotion -framework Foundation -framework Gini_iOS_SDK -framework Photos -compatibility_version 1 -current_version 1 -Xlinker -dependency_info -Xlinker /Users/agrosam/Library/Developer/Xcode/DerivedData/GiniDemo-dejipfpuyqllwcfijniaemxaqepo/Build/Intermediates.noindex/Pods.build/Debug-iphonesimulator/GiniVision.build/Objects-normal/x86_64/GiniVision_dependency_info.dat -o /Users/agrosam/Library/Developer/Xcode/DerivedData/GiniDemo-dejipfpuyqllwcfijniaemxaqepo/Build/Products/Debug-iphonesimulator/GiniVision/GiniVision.framework/GiniVision

Undefined symbols for architecture x86_64:
  "_OBJC_CLASS_$_BFTask", referenced from:
      objc-class-ref in AccountingDocumentService.o
      objc-class-ref in DocumentService.o
      objc-class-ref in DocumentServiceProtocol.o
  "_OBJC_CLASS_$_BFCancellationTokenSource", referenced from:
      objc-class-ref in AccountingDocumentService.o
      objc-class-ref in DocumentService.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

Reply:
Then, I created a vanilla Gini Demo with the following Podfile:

```
source 'https://github.com/gini/gini-podspecs.git'
source 'https://github.com/CocoaPods/Specs.git'
platform :ios, '9.0'


target 'GiniVanillaDemo' do
    use_frameworks!

    pod 'GiniVision'
    pod 'GiniVision/Networking'

end
``` 

Starting build, the linker issued the following error:

```

Showing All Messages
Ld /Users/agrosam/Library/Developer/Xcode/DerivedData/GiniVanillaDemo-batlbfhynykgzudkktgadoqkjdkh/Build/Products/Debug-iphonesimulator/GiniVision/GiniVision.framework/GiniVision normal x86_64 (in target: GiniVision)
    cd /Users/agrosam/Develop/GiniVanillaDemo/Pods
    export IPHONEOS_DEPLOYMENT_TARGET=9.0
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -arch x86_64 -dynamiclib -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator12.1.sdk -L/Users/agrosam/Library/Developer/Xcode/DerivedData/GiniVanillaDemo-batlbfhynykgzudkktgadoqkjdkh/Build/Products/Debug-iphonesimulator/GiniVision -F/Users/agrosam/Library/Developer/Xcode/DerivedData/GiniVanillaDemo-batlbfhynykgzudkktgadoqkjdkh/Build/Products/Debug-iphonesimulator/GiniVision -F/Users/agrosam/Library/Developer/Xcode/DerivedData/GiniVanillaDemo-batlbfhynykgzudkktgadoqkjdkh/Build/Products/Debug-iphonesimulator/Bolts -F/Users/agrosam/Library/Developer/Xcode/DerivedData/GiniVanillaDemo-batlbfhynykgzudkktgadoqkjdkh/Build/Products/Debug-iphonesimulator/Gini-iOS-SDK -filelist /Users/agrosam/Library/Developer/Xcode/DerivedData/GiniVanillaDemo-batlbfhynykgzudkktgadoqkjdkh/Build/Intermediates.noindex/Pods.build/Debug-iphonesimulator/GiniVision.build/Objects-normal/x86_64/GiniVision.LinkFileList -install_name @rpath/GiniVision.framework/GiniVision -Xlinker -rpath -Xlinker @executable_path/Frameworks -Xlinker -rpath -Xlinker @loader_path/Frameworks -mios-simulator-version-min=9.0 -dead_strip -Xlinker -object_path_lto -Xlinker /Users/agrosam/Library/Developer/Xcode/DerivedData/GiniVanillaDemo-batlbfhynykgzudkktgadoqkjdkh/Build/Intermediates.noindex/Pods.build/Debug-iphonesimulator/GiniVision.build/Objects-normal/x86_64/GiniVision_lto.o -Xlinker -export_dynamic -Xlinker -no_deduplicate -Xlinker -objc_abi_version -Xlinker 2 -fobjc-arc -fobjc-link-runtime -L/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/iphonesimulator -Xlinker -add_ast_path -Xlinker /Users/agrosam/Library/Developer/Xcode/DerivedData/GiniVanillaDemo-batlbfhynykgzudkktgadoqkjdkh/Build/Intermediates.noindex/Pods.build/Debug-iphonesimulator/GiniVision.build/Objects-normal/x86_64/GiniVision.swiftmodule -framework AVFoundation -framework CoreMotion -framework Foundation -framework Gini_iOS_SDK -framework Photos -compatibility_version 1 -current_version 1 -Xlinker -dependency_info -Xlinker /Users/agrosam/Library/Developer/Xcode/DerivedData/GiniVanillaDemo-batlbfhynykgzudkktgadoqkjdkh/Build/Intermediates.noindex/Pods.build/Debug-iphonesimulator/GiniVision.build/Objects-normal/x86_64/GiniVision_dependency_info.dat -o /Users/agrosam/Library/Developer/Xcode/DerivedData/GiniVanillaDemo-batlbfhynykgzudkktgadoqkjdkh/Build/Products/Debug-iphonesimulator/GiniVision/GiniVision.framework/GiniVision

Undefined symbols for architecture x86_64:
  "_OBJC_CLASS_$_BFTask", referenced from:
      objc-class-ref in AccountingDocumentService.o
      objc-class-ref in DocumentService.o
      objc-class-ref in DocumentServiceProtocol.o
  "_OBJC_CLASS_$_BFCancellationTokenSource", referenced from:
      objc-class-ref in AccountingDocumentService.o
      objc-class-ref in DocumentService.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

```

Reply:
Then it's not an issue in your project. Could you try to remove the Derive Data? 
`rm -rf ~/Library/Developer/Xcode/DerivedData`

Reply:
I promise, I did this already several times. :)

Reply:
But let me try also to delete the modul cache ....

Reply:
I guessed so, but it's strange that it works in several of our machines and haven't had any issue before :'(. Try removing Cocoapods cache as well: `pod cache clean --all`

Reply:
OK, since the error occurs intermittent, I tried several times. At least once, I got the linker error again. Sometimes it links successfully, but I believe it should not. ;)

Reply:
Please note, that we didn't have an issue with the patch. The changes in the Podspec are sane, and it only affects the linker in so far, that is causes Bolts Framework to be specified as dependency. The explicit import directive is correct, but not strictly required to have a successful build (which IMHO the compiler should check and issue a waring if it is omitted, but alas ...).

Reply:
OK, I try this too: pod cache clean --all 

Here is my MacBookPro spec: 
  Model Name:	MacBook Pro
  Model Identifier:	MacBookPro15,1
  Processor Name:	Intel Core i9
  Processor Speed:	2,9 GHz
  Number of Processors:	1
  Total Number of Cores:	6
  L2 Cache (per Core):	256 KB
  L3 Cache:	12 MB
  Memory:	32 GB

That's a decent one. It can compile a lot in parallel ...  just give some other pointer to a potential cause.

Reply:
OK, `pd cache clean --all` didn't help. Building directly after the command failed again. Performing clean cache, then pod update, then build, failed too.

Reply:
You have the `$(inherited)` attributed for the field OTHER LINKER FLAGS in your build settings, right?

![](https://i.stack.imgur.com/BbaLd.png)

Sorry if we are requiring you to check more things, but we would like to make sure that there is nothing wrong before changing anything, since it has been working flawlessly in several projects so far.

Reply:
The Other Linker Flags in the "Gini Vanilla Demo" for the "GiniVanillaDemo" target is as below:

```
OTHER_LDFLAGS = $(inherited) 
-framework "AVFoundation" 
-framework "Bolts" 
-framework "CoreMotion" 
-framework "GiniVision" 
-framework "Gini_iOS_SDK" 
-framework "Photos"
```

Please note though, that these linker options are effective only when linking the _final binary_ (the App executable). The error occurs _before_ this options affects ld, namely when linking the `GiniVision" dylib for the GiniVision Framework. 

In the transcripts above you can see all options when linking the GiniVision framework. The corresponding GiniVision Project has these Options set:

<img width="909" alt="screenshot 2019-02-08 at 09 01 13" src="https://user-images.githubusercontent.com/1564701/52465798-454d1a80-2b80-11e9-9db4-8edea4e1b652.png">

That is, none.

Reply:
We have been able to replicate the issue 🙏 , which only happens in the newest versions of _Cocoapods_. They have rewritten how they generate the build settings, being not possible to link the inner dependencies of subpecs automatically.
 
Comparing the build settings generated for the `GiniVision` library between different versions of _Cocoapods_, we were able to see what is missing in the newest versions:
`FRAMEWORK_SEARCH_PATHS = $(inherited) "${PODS_CONFIGURATION_BUILD_DIR}/Bolts" "${PODS_CONFIGURATION_BUILD_DIR}/Gini-iOS-SDK"`.

Thank you @couchdeveloper for being the first finding the issue and trying to find the root of it 👍. We'll test the changes proposed in this PR and if everything works, we will merge it. Stay tuned!

Reply:
Thank you very much, kikettas!  :).  I'm glad you were able to reproduce the issue and that you had that a great deal of patience to investigate it. :)  👍

Reply:
![](https://media.giphy.com/media/ui1hpJSyBDWlG/giphy.gif)